app/src/main/java/org/standardnotes/notes/comms/SyncManager.kt

```diff
@@ -1,55 +1,115 @@
 package org.standardnotes.notes.comms
 
+import android.os.Handler
 import org.standardnotes.notes.SApplication
 import org.standardnotes.notes.comms.data.Note
 import org.standardnotes.notes.comms.data.SyncItems
 import org.standardnotes.notes.comms.data.UploadSyncItems
 import retrofit2.Call
 import retrofit2.Callback
 import retrofit2.Response
+import java.lang.ref.WeakReference
 
 object SyncManager {
 
     interface SyncListener {
+        fun onSyncStarted()
+
         fun onSyncCompleted(notes: List<Note>)
 
         fun onSyncFailed()
     }
 
-    private val syncListeners: MutableList<SyncListener> = mutableListOf()
+    // TODO should we use a service instead?
+    private val SYNC_INTERVAL = 30000L
+    private val syncHandler: Handler = Handler()
+    private val syncRunnable: Runnable = object : Runnable {
+        override fun run() {
+            SyncManager.sync()
+            syncHandler.postDelayed(this, SYNC_INTERVAL)
+        }
+    }
+
+    private var syncing: Boolean = false
+    private val syncListeners: MutableList<WeakReference<SyncListener>> = mutableListOf()
 
     fun subscribe(listener: SyncListener) {
-        if (!syncListeners.contains(listener)) {
-            syncListeners.add(listener)
-        }
+        syncListeners
+                .filter { it.get() == listener }
+                .forEach { return }
+        syncListeners.add(WeakReference<SyncListener>(listener))
     }
 
     fun unsubscribe(listener: SyncListener) {
         syncListeners
-                .filter { listener == it }
+                .filter { listener == it.get() }
                 .forEach { syncListeners.remove(it) }
     }
 
     fun sync() {
+
+        if (syncing) {
+            return
+        }
+
+        syncing = true
+
+        var iter = syncListeners.iterator()
+        while (iter.hasNext()) {
+            val listening = iter.next()
+            if (listening.get() == null) {
```

Comment:
Will do. My initial thought was since it's using a weakreference, we could simplify it by doing away with unsubscribe altogether.  Once the fragment/activity is destroyed it would be free for garbage collection, and the weakreference here would be null. But that feels a bit dirty to me.